Sources/StreamChatUI/Utils/AudioPlayer/AudioPlayerObserving.swift

```diff
@@ -0,0 +1,132 @@
+    /// time-related metadata updates.
+    private var periodicTimeObservationToken: Any?
+
+    /// This block is called on deinit to remove any registered periodicTimeObserver
+    ///
+    /// According to the documentation we need to remove the periodicTimeObserver if we had
+    /// registered one.
+    /// https://developer.apple.com/documentation/avfoundation/avplayer/1385829-addperiodictimeobserverforinterv#return_value
+    private var periodicTimeObservationCancellationBlock: (() -> Void)?
+
+    /// An observer token that is used to reference the NotificationCenter registration, that is being used
+    /// to get notifications when the player's playback has been stopped.
+    private var stoppedPlaybackObservationToken: Any?
+
+    /// The notificationCenter on which the ``playbackFinishedObserver`` will be registered on
+    private let notificationCenter: NotificationCenter
+
+    // MARK: - Lifecycle
+
+    public init(
+        notificationCenter: NotificationCenter = .default
+    ) {
+        self.notificationCenter = notificationCenter
+    }
+
+    deinit {
+        timeControlStatusObserver?.invalidate()
+        periodicTimeObservationCancellationBlock?()
+        stoppedPlaybackObservationToken.map { notificationCenter.removeObserver($0) }
+    }
+
+    // MARK: - AudioPlayerObserving
+
+    open func addTimeControlStatusObserver(
+        _ player: AVPlayer,
+        using block: @escaping (AVPlayer.TimeControlStatus?) -> Void
+    ) {
+        timeControlStatusObserver = player.observe(
+            \.timeControlStatus,
+            changeHandler: { _, change in block(change.newValue) }
+        )
+    }
+
+    open func addPeriodicTimeObserver(
+        _ player: AVPlayer,
+        forInterval interval: CMTime,
+        queue: DispatchQueue?,
+        using block: @escaping () -> Void
+    ) {
+        periodicTimeObservationToken = player.addPeriodicTimeObserver(
+            forInterval: interval,
+            queue: queue,
+            using: { _ in block() }
+        )
+        periodicTimeObservationCancellationBlock = { [weak player, periodicTimeObservationToken] in
+            guard
+                let player = player,
+                let periodicTimeObservationToken = periodicTimeObservationToken
+            else {
+                return
+            }
+            player.removeTimeObserver(periodicTimeObservationToken)
+        }
+    }
+
+    open func addStoppedPlaybackObserver(
+        queue: OperationQueue?,
+        using block: @escaping (AVPlayerItem) -> Void
+    ) {
+        stoppedPlaybackObservationToken = notificationCenter.addObserver(
+            forName: NSNotification.Name.AVPlayerItemDidPlayToEndTime,
+            object: nil,
+            queue: queue
+        ) { notification in
+            guard
+                let playerItem = notification.object as? AVPlayerItem
+            else {
+                return
```

Comment:
Is this case possible? Would there be any implications to the UI layer?

Reply:
As this notification can be fired from anyone, including AVPlayer, I would say it's possible. I don't foresee any UI implications as if it's not our player firing it, we will ignore it.